### PR TITLE
[opt] Constant folding for unary ops on host

### PR DESCRIPTION
**Pull request by ailzhang**
Note this PR might cause some precision differences since we used to do constant folding on device while it's now on the host. 

Issue: #

### CODE REVIEW

File: taichi/transforms/constant_fold.cpp [hunk 1]
```diff
@@ -176,30 +149,76 @@ class ConstantFold : public BasicStmtVisitor {
...
+      return;
     }
+    const auto dt = operand->val.dt;
+    if (!is_good_type(dt))
+      return;
+    const auto dst_type = stmt->ret_type;
+    switch (stmt->op_type) {
+#define HANDLE_REAL_AND_INTEGRAL_UNARY(OP_TYPE, OP_CPP)                     \
+  case UnaryOpType::OP_TYPE: {                                              \
+    if (dt->is_primitive(PrimitiveTypeID::f32) ||                           \
```
Comment (lin-hitonami) [inline taichi/transforms/constant_fold.cpp:165]: We can use `is_real`, `is_signed`, and `is_unsigned` functions here

File: taichi/transforms/constant_fold.cpp [hunk 2]
```diff
@@ -176,30 +149,76 @@ class ConstantFold : public BasicStmtVisitor {
...
+        dt->is_primitive(PrimitiveTypeID::f64)) {                           \
+      auto res = TypedConstant(dst_type, OP_CPP(operand->val.val_float())); \
+      insert_and_erase(stmt, res);                                          \
+    } else if (dt->is_primitive(PrimitiveTypeID::i32) ||                    \
```
Comment (lin-hitonami) [inline taichi/transforms/constant_fold.cpp:169]: Don't we support 8-bit and 16-bit data types?

Comment (ailzhang) [reply taichi/transforms/constant_fold.cpp:169]: @lin-hitonami we didn't do constant folding for 8-bit and 16-bits before (see `is_good_type` function ;) We can definitely add those as well but I'd prefer to add them in a separate PR.

File: taichi/transforms/constant_fold.cpp (hunk 1 above)
Comment (ailzhang) [reply taichi/transforms/constant_fold.cpp:165]: Yup the reason that we list types explicitly here is to align with `is_good_type` but we can update these for sure!

File: taichi/transforms/constant_fold.cpp [hunk 3]
```diff
@@ -176,30 +149,76 @@ class ConstantFold : public BasicStmtVisitor {
...
+      HANDLE_REAL_AND_INTEGRAL_UNARY(acos, std::acos)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(tan, std::tan)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(tanh, std::tanh)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(log, std::log)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(exp, std::exp)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(cast_value, )
+      HANDLE_REAL_AND_INTEGRAL_UNARY(rsqrt, 1.0 / std::sqrt)
+#undef HANDLE_REAL_AND_INTEGRAL_UNARY
+
+#define HANDLE_INTEGRAL_UNARY(OP_TYPE, OP_CPP)                             \
```
Comment (jim19930609) [inline taichi/transforms/constant_fold.cpp:199]: Correct me if were wrong, but looks like `HANDLE_INTEGRAL_UNARY` is a subset of `HANDLE_REAL_AND_INTEGRAL_UNARY`? Why don't we just put `HANDLE_REAL_AND_INTEGRAL_UNARY` here?

**Review by jim19930609 — APPROVED** [review]

File: taichi/transforms/constant_fold.cpp (hunk 3 above)
Comment (ailzhang) [reply taichi/transforms/constant_fold.cpp:199]: @jim19930609 the main reason is we actually don't support bit_not/logical_not on real types so I didn't want to fake the support here. (but yea we should have done the type check beforehand...

Comment (ailzhang) [reply taichi/transforms/constant_fold.cpp:199]: I'm going to send a separate PR for the type check & error message ;)

**Pull request merged**